javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -348,95 +348,136 @@ object Math {
     }
   }
 
+  private def intOverflow(): Nothing =
+    throw new ArithmeticException("Integer overflow")
+
+  private def longOverflow(): Nothing =
+    throw new ArithmeticException("Long overflow")
+
+  @inline
   def addExact(a: scala.Int, b: scala.Int): scala.Int = {
+    // Hacker's Delight, Section 2-13
     val res = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Integer overflow")
+    if (((res ^ a) & (res ^ b)) < 0)
+      intOverflow()
+    res
   }
 
+  @inline
   def addExact(a: scala.Long, b: scala.Long): scala.Long = {
+    /* Hacker's Delight, Section 2-13
+     * With RuntimeLong, the computations of all the lo words in the overflow
+     * check are dead-code eliminated.
+     */
     val res = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Long overflow")
+    if (((res ^ a) & (res ^ b)) < 0L)
+      longOverflow()
+    res
   }
 
+  @inline
   def subtractExact(a: scala.Int, b: scala.Int): scala.Int = {
+    // Hacker's Delight, Section 2-13
     val res = a - b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b > 0)) res
-    else throw new ArithmeticException("Integer overflow")
+    if (((a ^ b) & (res ^ a)) < 0)
+      intOverflow()
+    res
   }
 
+  @inline
   def subtractExact(a: scala.Long, b: scala.Long): scala.Long = {
+    /* Hacker's Delight, Section 2-13
+     * With RuntimeLong, the computations of all the lo words in the overflow
+     * check are dead-code eliminated.
+     */
     val res = a - b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b > 0)) res
-    else throw new ArithmeticException("Long overflow")
+    if (((a ^ b) & (res ^ a)) < 0L)
+      longOverflow()
+    res
```

Comment:
I do not understand how the overflow checks for addition / subtraction up to here in the file relate to Hacker's Delight. But I understand that they essentially do the same as before, just with bitwise operations. (so no additional comments required IMO, but maybe remove the hacker's delight comments?).

Reply:
I elaborated the comments to point more specifically to the formulas we use. But you're right, it's basically applying De Morgan to the negation of our previous formulas.